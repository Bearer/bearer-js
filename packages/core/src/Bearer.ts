import fbemitter, { EventSubscription } from 'fbemitter'
import postRobot from 'post-robot'

import BearerConfig from './BearerConfig'
import Events from './EventNames'

const BEARER_WINDOW_INSTANCE_KEY = 'BEARER_INSTANCE'
const BEARER_EMITTER = 'BEARER_EMITTER'
const BEARER_WINDOW_KEY = 'BEARER'
const BEARER_CONFIG_KEY = 'BEARER_CONFIG'
const IFRAME_NAME = 'BEARER-IFRAME'

type TAuthorizationPayload = {
  scenarioId: string
  authIdentifier?: string
}

class Bearer {
  private static set _instance(bearerInstance: Bearer) {
    if (window[BEARER_WINDOW_INSTANCE_KEY]) {
      console.warn('[BEARER]', 'Replacing bearer instance')
    }
    window[BEARER_WINDOW_INSTANCE_KEY] = bearerInstance
  }

  private static get _instance(): Bearer | undefined {
    return window[BEARER_WINDOW_INSTANCE_KEY]
  }

  public static get emitter(): fbemitter.EventEmitter {
    if (!window[BEARER_EMITTER]) {
      window[BEARER_EMITTER] = new fbemitter.EventEmitter()
    }
    return window[BEARER_EMITTER]
  }

  public static get instance(): Bearer {
    return window[BEARER_WINDOW_INSTANCE_KEY] || this.init({})
  }

  public static get config(): BearerConfig {
    return this.instance.bearerConfig
  }

  static get version() {
    return 'LIB_VERSION'
  }

  get bearerConfig(): BearerConfig {
    return window[BEARER_WINDOW_KEY]
  }

  set bearerConfig(config: BearerConfig) {
    window[BEARER_WINDOW_KEY] = config
  }

  get maybeInitialized(): Promise<boolean> {
    if (!this.isSessionInitialized) {
      console.warn('[BEARER]', 'Waiting Bearer to be initialized')
    }
    return this._maybeInitialized
  }

  set maybeInitialized(promise) {
    this._maybeInitialized = promise
  }
  public static init(config: any = {}): Bearer {
    if (this._instance) {
      console.warn('One instance is already configured, reaplacing it')
    }
    this._instance = new Bearer({ ...config, ...(window[BEARER_CONFIG_KEY] || {}) })
    return this._instance
  }

  static onAuthorized = (scenarioId: string, callback: (authorize: boolean) => void): EventSubscription => {
    console.log('[BEARER]', 'register onAuthorized', scenarioId)
    return Bearer.emitter.addListener(Events.AUTHORIZED, () => {
      // TODO : listen only for the scenarioId (+ setupId ?)
      console.log('[BEARER]', 'calling authorized listener', scenarioId)
      callback(true)
    })
  }

  static onRevoked = (scenarioId: string, callback: (authorize: boolean) => void): EventSubscription => {
    console.log('[BEARER]', 'register onRevoked', scenarioId)
    return Bearer.emitter.addListener(Events.REVOKED, () => {
      // TODO : listen only for the scenarioId (+ setupId ?)
      console.log('[BEARER]', 'calling revoke listener', scenarioId)
      callback(false)
    })
  }

  private iframe: HTMLIFrameElement
  private isSessionInitialized: boolean = false
  private allowIntegrationRequests: (initialize: true) => void
  private _maybeInitialized: Promise<boolean>

  constructor(args) {
    this.bearerConfig = new BearerConfig(args || {})
    console.info('[BEARER]', 'config initialized with', args)
    this.maybeInitialized = new Promise((resolve, _reject) => {
      this.allowIntegrationRequests = resolve
    })
    this.initSession()
  }

  authorized = (data: TAuthorizationPayload) => {
    console.log('[BEARER]', 'Bearer:emitting ', { event: Events.AUTHORIZED, ...data })
    Bearer.emitter.emit(Events.AUTHORIZED, data)
  }

  revoked = (data: TAuthorizationPayload) => {
    console.log('[BEARER]', 'Bearer:emitting ', { event: Events.REVOKED, ...data })
    Bearer.emitter.emit(Events.REVOKED, data)
  }

  hasAuthorized = (scenarioId): Promise<boolean> =>
    new Promise((resolve, reject) => {
      postRobot
        .send(this.iframe, Events.HAS_AUTHORIZED, {
          scenarioId,
          clientId: Bearer.config.clientId
        })
        .then(({ data, data: { authorized } }) => {
          console.log('[BEARER]', 'HAS_AUTHORIZED response', data)
          authorized ? resolve(true) : reject(false)
        })
        .catch(iframeError)
    })

  revokeAuthorization = (scenarioId: string): void => {
    postRobot
      .send(this.iframe, Events.REVOKE, {
        scenarioId,
        clientId: Bearer.config.clientId
      })
      .then(() => {
        console.log('[BEARER]', 'Signing out')
      })
      .catch(iframeError)
  }

  initSession() {
    if (window !== undefined && !document.querySelector(`#${IFRAME_NAME}`)) {
      postRobot.on(Events.SESSION_INITIALIZED, event => {
        this.sessionInitialized(event)
      })
      postRobot.on(Events.AUTHORIZED, this.authorized)
      postRobot.on(Events.REVOKED, this.revoked)

      this.iframe = document.createElement('iframe')
      this.iframe.src = `${this.bearerConfig.authorizationHost}v1/user/initialize`
      this.iframe.id = IFRAME_NAME
      this.iframe.width = '0'
      this.iframe.height = '0'
      this.iframe.frameBorder = '0'
      document.body.appendChild(this.iframe)
    }
  }

  askAuthorizations = ({ scenarioId, setupId }): boolean => {
    if (this.isSessionInitialized) {
      const AUTHORIZED_URL = `${Bearer.config.integrationHost}v1/auth/${scenarioId}?setupId=${setupId}`
      window.open(AUTHORIZED_URL, '', 'resizable,scrollbars,status,centerscreen=yes,width=500,height=600')
      return true
    }
    return false
  }

  private sessionInitialized(_event) {
    console.log('[BEARER]', 'session initialized')
    this.isSessionInitialized = true
    this.allowIntegrationRequests(true)
  }
}

function iframeError(e) {
  console.error('[BEARER]', 'Error contacting iframe', e)
}

export default Bearer
