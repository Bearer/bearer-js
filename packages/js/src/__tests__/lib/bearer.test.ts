import Bearer, { findElements } from '../../lib/bearer'

const clientId = 'a-client-id'

describe('main', () => {
  it('exports a Bearer class', () => {
    expect(Bearer).toBeTruthy()
  })

  describe('constructor', () => {
    describe('readyState ', () => {
      it('register a listerer on dom DOMContentLoaded if loading', () => {
        Object.defineProperty(document, 'readyState', {
          configurable: true,
          get() {
            return 'loading'
          }
        })
        document.addEventListener = jest.fn()

        new Bearer(clientId)

        expect(document.addEventListener).toHaveBeenCalledWith('DOMContentLoaded', expect.any(Function))
      })

      it('does not register listerner if complete', () => {
        Object.defineProperty(document, 'readyState', {
          get() {
            return 'complete'
          }
        })

        document.addEventListener = jest.fn()

        new Bearer(clientId)

        expect(document.addEventListener).not.toHaveBeenCalled()
      })

      it('does not register listerner if interactive', () => {
        Object.defineProperty(document, 'readyState', {
          get() {
            return 'interactive'
          }
        })

        document.addEventListener = jest.fn()

        new Bearer(clientId)

        expect(document.addEventListener).not.toHaveBeenCalled()
      })
    })
  })

  describe('fetch request', () => {
    it.skip('returns payload if success', () => {
      expect(false).toBeTruthy()
    })
  })
  describe('function request', () => {
    it.skip('returns payload if success', () => {
      expect(false).toBeTruthy()
    })
  })
})

describe('findElements', () => {
  beforeAll(() => {
    document.body.innerHTML = `
      <bearer-something></bearer-something>
      <bearer-patrick></bearer-patrick>
      <spongebob-something></spongebob-something>
      <notmatching-anything></notmatching-anything>
    `
  })

  describe('matching elements from default', () => {
    it('returns lowercased matching tab names', () => {
      expect(findElements(document.querySelectorAll('*'))).toEqual(['bearer-something', 'bearer-patrick'])
    })
  })

  describe('matching elements bearer tag by default', () => {
    it('returns lowercased matching tab names', () => {
      expect(findElements(document.querySelectorAll('*'), /^spongebob-/i)).toEqual(['spongebob-something'])
    })
  })
})
