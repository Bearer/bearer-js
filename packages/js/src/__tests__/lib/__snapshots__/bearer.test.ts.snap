// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sd sd 1`] = `
<body>


  <bearer-something />


  <bearer-patrick />


  <spongebob-something />


  <notmatching-something />


</body>
`;
